### DISCUSSION TITLE
Does `scopeKey` function the same as `queryKey` in Tanstack Query?

### DISCUSSION THREAD

**Discussion by b-pmcg** [wevm/wagmi#2050] [Q&A] [answered]
I was reading up on how caching works and I came across [this post by tkdodo](https://tkdodo.eu/blog/practical-react-query#treat-the-query-key-like-a-dependency-array) that suggests leveraging variables in the queryKey to trigger a refetch.
>React Query will trigger a refetch whenever the query key changes. So when we pass a variable parameter to our queryFn, we almost always want to fetch data when that value changes. Instead of orchestrating complex effects to manually trigger a refetch, we can utilize the query key.

Does the scope key function in the same way? I know you can't pass an array to a `scopeKey` but you could use template literals.

**Comment by tmm** [comment, marked as answer]
`scopeKey` gets added to the existing internal query key. you could use it to trigger a refetch, but note that it will also trigger a refetch when any of the dependencies of the internal query key change (e.g. for `useEnsName` `address`).

**Reply by b-pmcg** [reply]
ok, with the `useContractReads` hook for example, does this mean a refetch be triggered if any of the `args`  in the `contracts` array change?

**Reply by tmm** [reply]
yes, it should

**Reply by ArianHamdi** [reply]
@tmm 

What if we need to update cache manually for a specefic contract ?
How can I find `queryKey` for each `useContractRead` or any option to provide `queryKey` to the hook.
In React Query I will call `queryClient.setQueryData` and pass queryKey and update that specefic item.

My use case : I need to update cache in `useContractEvent`  listener when new data comes.

**Reply by AkshatGiri** [reply]
@ArianHamdi Did you find a solution to this? I want to optimistically update the query data when the transaction is sent ( even before it's confirmed ).

**Reply by ArianHamdi** [reply]
I used `wagmi/core` and manually integrated it with `react-query`, so in this case I am able to determine `queryKey`.
